### DISCUSSION TITLE
Can't deserialize with converter provider and kotlinx.serialization

### DISCUSSION THREAD

**Discussion by nucle** [jOOQ/jOOQ#15176] [Q&A] [answered]
Hi,
I try to deserialize a JSONB column with a converter provider and  [kotlinx.serialization](https://github.com/Kotlin/kotlinx.serialization). I get the following exception from
the converter:

> Serializer for class 'Class' is not found.
> Please ensure that class is marked as '@Serializable' and that the serialization compiler plugin is applied.

My data classes are defined as serializable and the [compiler plugin](https://plugins.gradle.org/plugin/org.jetbrains.kotlin.plugin.serialization) is applied.

Can anyone help, please?

**Select JSONB column**
```kotlin
    fun configByApiKey(apiKey: String): Config? = dslContext.select(
        CONFIGS.PAGE_FLOW, // JSONB
       // some other columns
    )
        .from(CONFIGS)
        .where(CONFIGS.API_KEY.eq(apiKey))
        .fetchOneInto(Config::class.java)
```
**Select result: Array of objects**
+--------------------------------------------------+
 |page_flow                                                                    |
+--------------------------------------------------+
|[{"name":"locations","order":0},{"name":"servic...   |
+--------------------------------------------------+

**Data classes:**
```kotlin
@Serializable
data class Config(
    @SerialName("pageFlow")
    val pageFlow: List<PageFlow>,
)

@Serializable
class PageFlow(val name: String, val order: Int)
```

**ConverterProvider**
```kotlin
internal class JSONConverterProvider(private val json: Json) : ConverterProvider {
    private val delegate: ConverterProvider = DefaultConverterProvider()

    @OptIn(InternalSerializationApi::class)
    override fun <T, U> provide(tType: Class<T>, uType: Class<U>): Converter<T, U>? {

        // Our specialised implementation can convert from JSON (optionally, add JSONB, too)
        return if (tType == JSONB::class.java) {
            Converter.ofNullable(tType as Class<T>, uType as Class<U>,
                { t : T ->
                    try {
                        val data = (t as JSONB).data()

// Serializer for class 'Class' is not found.
// Please ensure that class is marked as '@Serializable' and that the serialization compiler plugin is applied.

                        val deserialized = json.decodeFromString(uType::class.serializer(), data )
                        return@ofNullable deserialized as U
                    } catch (e: Exception) {
                        throw DataTypeException("JSON mapping error", e)
                    }
                }
            ) { u: U ->
                try {
                    return@ofNullable null
                } catch (e: Exception) {
                    throw DataTypeException("JSON mapping error", e)
                }
            }
        } else delegate.provide<T, U>(tType, uType)
    }
}
```

**Comment by lukaseder** [comment, marked as answer]
It seems you cannot use `uType::class`. The function you're using there is `KClass<T>.serializer()`:
https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-core/kotlinx.serialization/serializer.html

But `uType::class` returns `KClass<java.lang.Class<U>>`, not `KClass<U>`.

Here's how you can turn one into the other: https://stackoverflow.com/q/55369415/521799

**Comment by nucle** [comment]
Hi @lukaseder,
thank you for your fast response and help.

[ Serializer API ](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-core/kotlinx.serialization/serializer.html)
````
json.decodeFromString(serializer(uType), data) as U
````

Thank You!

One question:
Is it possible to access the parsed data? 
<img width="341" alt="image" src="https://github.com/jOOQ/jOOQ/assets/2041897/57737b25-30d9-4416-9007-85bfa11a2139">



BR,
Chris

**Reply by lukaseder** [reply]
Great to hear, thanks for documenting the solution here!

**Reply by nucle** [reply]
Hi, 
here is a solution for a custom converter:

In gradle we need to define our forced types:
````
...gradle
ForcedType()
        .withUserType("at.knusperleicht.domain.Style")
        .withConverter("at.knusperleicht.JsonConverter(Style::class.java)")
        .withIncludeExpression(".*\\.style"),
...
````

Our kotlinx converter looks like this:
````kotlin
import kotlinx.serialization.json.Json
import kotlinx.serialization.serializer
import org.jooq.JSONB
import org.jooq.impl.AbstractConverter

class JsonConverter<U : Any>(
    toClass: Class<U>,
    private val json: Json = Json
) : AbstractConverter<JSONB, U>(JSONB::class.java, toClass ) {

    override fun from(databaseObject: JSONB): U {
        return json.decodeFromString(serializer(toType()), databaseObject.data()) as U
    }

    override fun to(userObject: U): JSONB {
        return JSONB.jsonb(json.encodeToString(serializer(toType()), userObject))
    }
}
````

BR,
Chris
